### DISCUSSION TITLE
Add ability to set material and shadow blend modes separate from one another

### DISCUSSION THREAD

**Discussion by catboy-catfish** [godotengine/godot-proposals#9261] [3D]
![image](https://github.com/godotengine/godot-proposals/assets/94094139/b6b5090e-f370-4d48-9f83-b111238fb3c4)

I just had this idea for creating realistic(ish) transparent surfaces with transparent shadows.
The idea is that we use normal alpha blending for the surface itself, but use hashing and/or dithering for its shadows, however I don't think I can pull this off in Godot 4.21 Stable right now. I _could_ just make a fork and try to implement this myself, but I'm not smart yet.

![image](https://github.com/godotengine/godot-proposals/assets/94094139/500336e2-eb4b-409a-aba0-7d2da663469e)
Right now I have this custom glass shader I did which uses Depth Prepass for the shadows. The surface looks great. The shadows look fine but not as good as they could be.

![image](https://github.com/godotengine/godot-proposals/assets/94094139/2982eb2d-e772-4275-be81-5114836166e4)
Best I can use is Hash blending, which does the shadows beautifully but it makes the surface look terrible.

![image](https://github.com/godotengine/godot-proposals/assets/94094139/20cfb647-a0f3-47fd-9e19-31519c45e6b4)
TAA can reduce the grain/artifacts, but not completely so. It also creates new ghosting artifacts due to its temporal nature.

This is my feature request: add an ability to set the shadow's blending mode separately from the surface's. Blender EEVEE is capable of doing this, and I've heard Unity takes a similar approach. Right now I'm using version 4.2.1 Stable and I'll likely stick to that until a stable version of something like 4.3 is released. Maybe you could add this to 4.3 stable or in the meantime.

Thank you

**Comment by Calinou** [comment]
- Related to https://github.com/godotengine/godot-proposals/issues/3276.

This is already possible by using two MeshInstance3D nodes with the same mesh, at the same position:

- One that has its Cast Shadows property set to **Off** (the display mesh).
- One that has its Cast Shadow property set to **Shadows Only** (the shadow mesh).

In the shadow pass (`AT_SHADOW_PASS`), we could have alpha hashing be used when GeometryInstance3D Transparency property is used when this value is greater than `0.0`, since this doesn't disable shadow casting right now. In comparison, setting the transparency mode to **Alpha** will always disable shadow casting, so you can't handle `AT_SHADOw_PASS` in the shader.

> Maybe you could add this to 4.3 stable or in the meantime.

4.3 is nearing feature freeze, so any new features that need a significant amount of work like this will be for 4.4 at the earliest.
